geowebcache/diskquota/core/src/main/java/org/geowebcache/diskquota/storage/TilePageCalculator.java

```diff
@@ -99,48 +111,59 @@ public BigInteger getTilesPerPage(TileSet tileSet, int zoomLevel) {
     }
 
     public Set<TileSet> getTileSetsFor(final String layerName) {
-
-        TileLayer tileLayer;
-        try {
-            tileLayer = tld.getTileLayer(layerName);
-        } catch (GeoWebCacheException e) {
-            throw new IllegalArgumentException(e);
-        }
-
-        final Collection<GridSubset> gridSubSets = new ArrayList<GridSubset>();
-        for (String gridSetId : tileLayer.getGridSubsets()) {
-            GridSubset subset = tileLayer.getGridSubset(gridSetId);
-            gridSubSets.add(subset);
-        }
-        final List<MimeType> mimeTypes = tileLayer.getMimeTypes();
-
-        return getTileSetsFor(layerName, gridSubSets, mimeTypes);
+        return getTileSetsFor(layerName, Optional.empty(), Optional.empty(), Optional.empty());
     }
 
     public Set<TileSet> getTileSetsFor(final String layerName,
             final Collection<GridSubset> gridSubSets, final List<MimeType> mimeTypes) {
-        Set<TileSet> layerTileSets = new HashSet<TileSet>();
-
-        String gridsetId;
-        String blobFormat;
-        String parametersId = null;
-        // TODO: create one TileSet per parametersId (somehow...)
-
-        TileSet tileSet;
-        for (GridSubset gridSubset : gridSubSets) {
-            gridsetId = gridSubset.getName();
-
-            for (MimeType mime : mimeTypes) {
-                blobFormat = mime.getFormat();
-
-                tileSet = new TileSet(layerName, gridsetId, blobFormat, parametersId);
-                layerTileSets.add(tileSet);
+        
+        return getTileSetsFor(layerName, Optional.of(gridSubSets), Optional.of(mimeTypes), Optional.empty());
+    }
+    
+    public Set<TileSet> getTileSetsFor(
+            final String layerName, 
+            final Optional<? extends Collection<GridSubset>> gridSubsets, 
+            final Optional<? extends Collection<MimeType>> mimeTypes,
+            final Optional<? extends Collection<String>> parameterIds) {
+        
+        // If we don't have gridsets or mime types, we need to look up the layer object
+        final TileLayer tileLayer;
+        if(!(gridSubsets.isPresent() && mimeTypes.isPresent())) {
+            try {
+                tileLayer = tld.getTileLayer(layerName);
+            } catch (GeoWebCacheException e) {
+                throw new IllegalArgumentException(e);
             }
+        } else {
+            tileLayer = null;
         }
-
-        return layerTileSets;
+        
+        // Loop over the gridset names, getting them from the layer if they were not provided
+        return gridSubsets
```

Comment:
Another bit that has too much density/nesting imho, splitting it a bit would make it more readable.